Merge Split algorithm added

kelcyno opened the pull request:
The addition in this pull request is a post processing algorithm for the tobac tracking. This file contains the merge_split function to address merging/splitting cells in the tobac track, as well as a quick compression function, and a function to standardize the track dataset in order to have all tracking, feature, and segmentation mask datasets in one xarray dataset. 
An example of these functions is given below:  

d = merge_split(Track)
ds = standardize_track_dataset(Track, refl_mask, data['ProjectionCoordinateSystem'])
 both_ds = xr.merge([ds, d],compat ='override')
 both_ds = compress_all(both_ds)
 both_ds.to_netcdf(os.path.join(savedir,'Track_features_merges.nc'))

kelcyno commented:
@freemansw1 Nothing was going right on my end so I just started fresh. Let me know if there's anything wonky.

JuliaKukulies commented:
@kelcyno Thank you, I will also start to have a look your PR now. Can you quickly fix the formatting of `merge_split.py` ?

JuliaKukulies reviewed (changes requested):
Thank you so much for this PR @kelcyno!  I am still figuring out some details in your code, because I am not sure I completely understand what the `merge_split()` function does at all steps.  But overall it looks good and I have tested your functions with some of my data which worked fine (except for the data standardization). Please have a look at my comments and questions below which we can solve together before I approve this!

freemansw1 commented:
Sorry for my delay on this, I should be able to review next week after my defense. @kelcyno do you want me to review ASAP or are there changes to be made from @JuliaKukulies's review?

kelcyno commented:
Sean, I still have some edits from Julia to work on. It can wait until after your defense.

On Jun 22, 2022, at 6:24 AM, Sean Freeman ***@***.***> wrote:

﻿
This email originated outside TTU. Please exercise caution<https://askit.ttu.edu/phishing>!


Sorry for my delay on this, I should be able to review next week after my defense. @kelcyno<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkelcyno&data=05%7C01%7Ckelcy.brunner%40ttu.edu%7C4fbb2373e12c47c9d9f408da53feb979%7C178a51bf8b2049ffb65556245d5c173c%7C0%7C0%7C637914650787197382%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Eno%2FbEl%2BUQFJlTE78mUDVvavcIVNQevGk9MlYIsEPUI%3D&reserved=0> do you want me to review ASAP or are there changes to be made from @JuliaKukulies<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FJuliaKukulies&data=05%7C01%7Ckelcy.brunner%40ttu.edu%7C4fbb2373e12c47c9d9f408da53feb979%7C178a51bf8b2049ffb65556245d5c173c%7C0%7C0%7C637914650787197382%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=DqxTN%2F0q7KYLCOgm%2BCC9TU%2FSOzhEI6DLHZCkDhKZn%2BE%3D&reserved=0>'s review?

—
Reply to this email directly, view it on GitHub<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Ftobac-project%2Ftobac%2Fpull%2F136%23issuecomment-1162593375&data=05%7C01%7Ckelcy.brunner%40ttu.edu%7C4fbb2373e12c47c9d9f408da53feb979%7C178a51bf8b2049ffb65556245d5c173c%7C0%7C0%7C637914650787197382%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Wm83MQq2YvXUKgqBF70V8GIOl%2Bw21SteiwRJTysZlxQ%3D&reserved=0>, or unsubscribe<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAU7Z2UZVLL5YTSB6HLCRFQ3VQKBPBANCNFSM5YAZ4APQ&data=05%7C01%7Ckelcy.brunner%40ttu.edu%7C4fbb2373e12c47c9d9f408da53feb979%7C178a51bf8b2049ffb65556245d5c173c%7C0%7C0%7C637914650787197382%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=tPN1IjioM1324%2FvT8XtFlFX4suxkZZ4wW6E6eD%2FqlzM%3D&reserved=0>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

freemansw1 commented:
OK, I'll wait to review this until you have made the changes @kelcyno . When you do, we can re-request reviews from both @JuliaKukulies and myself.

One note, with only having glanced at the PR, if we need to add new requirements for this, we should either make them optional (add a check to see if the package can be imported when trying to use this function; if not throw an error and quit) or add them to the various requirements files.

kelcyno commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,345 @@
+#Tobac merge and split v0.1
+
+from geopy.distance import geodesic
```
Hi Julia, I uncommented the section you mentioned, adding the option to check if the lat/lon is available in tracks. Thus I left this import in place as is.

kelcyno commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,345 @@
+#Tobac merge and split v0.1
```
I updated the documentation for the sub modules and each function within the file - let me know if I did it correctly or you'd like additional documentation.

kelcyno commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,345 @@
+    
+
+def standardize_track_dataset(TrackedFeatures, Mask, Projection):
+    """ Combine a feature mask with the feature data table into a common dataset.
```
data['ProjectionCoordinateSystem'] does refer to the original input data - which in our instance is usually gridded radar data. I made this input optional, it is not necessary for the standardize_track_dataset merging.

kelcyno commented:
@freemansw1 @JuliaKukulies I updated the merge_split.py to comply with documentation and Julia's previous comments. Looking forward to more input/comments.

kelcyno commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,345 @@
+                        
+    Example:
+        d = merge_split(Track)
+       ds = standardize_track_dataset(Track, refl_mask, data['ProjectionCoordinateSystem'])
```
@JuliaKukulies I just missed this comment in my pushed changes. I'll update shortly with a clearer example. The input to standardized_track_dataset() is  the Tracked file (NOT the merge_split post processed tracks). The purpose of the standardized_track_dataset() is to merge the features, segmentation mask, and the linked tracks together. I then merge these in the example with the postprocessed merge_split tracks.

kelcyno commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,345 @@
+    d = d.set_coords(['feature_id','cell_id', 'track_id', 'storm_id']) 
+
+    assert len(track_id) == len(track_parent_storm_id)
+    assert len(cell_id) == len(cell_parent_track_id)
```
I will make these changes and add to the test dataset - again, sorry I missed this comment.

kelcyno replied:
@freemansw1 @JuliaKukulies To add this to the tests folder I need to include a test dataset - I typically use gridded radar data. The test 2d/3d data in the testing.py does not create anything that merge_split.py would merge or split and is not useful within a test function. What are your thoughts here?

freemansw1 replied:
I wonder if we could build a "fake" pair of features that merge/split/etc. using some of the tools in `testing.py`. I'm going to explore this some as I get to reviewing/testing this. Hopefully the functions should be decently documented, but I added most of them, so if there are issues, let me know.

JuliaKukulies replied:
What is the status on the testing front ? :)

freemansw1 commented:
Thanks @kelcyno , I'll take a look at this today or tomorrow. Just so you know- the reason your PR is failing checks is that you need to run the `black` formatter on your new file. 

One question before I begin review- do you have documentation handy for this?

freemansw1 commented:
Just noting that I know you are still waiting on a review from me. It's been very busy and I haven't quite had the time to give reviewing this the attention it deserves. I should have time early next week, though. I do have some ideas both on the documentation front and the testing front though, and I'll post those when I post my review.

freemansw1 commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,477 @@
+        xarray dataset of merged Track and Segmentation Mask datasets with renamed variables.
+
+    """
+    feature_standard_names = {
```
I think it's possible that these would be better off in a JSON. That would allow us to pull from there to build the documentation, too. However, I don't think that's necessary for this PR.

freemansw1 commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,477 @@
+        if k in TrackedFeatures.variables.keys()
+    }
+
+    TrackedFeatures = TrackedFeatures.drop(["cell_parent_track_id"])
```
Why does this get dropped? I don't think this column should normally appear in the track dataframe/xarray

freemansw1 commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,477 @@
+        g.add_edge(a_names[i], b_names[i], weight=dist[i])
+
+    tree = minimum_spanning_edges(g)
+    tree_list = list(minimum_spanning_edges(g))
```
Is there a reason `minimum_spanning_edges` is run twice? can `tree_list` just be `list(tree)`?

kelcyno replied:
Reduced it as suggested.

freemansw1 commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,477 @@
+                track_id[np.nanmax(np.unique(temp))] = list(np.unique(temp))
+
+    storm_id = [0]  # default because we don't track larger storm systems *yet*
+    print("found storm id")
```
As we discussed, I am inclined to remove the storm object for now. @JuliaKukulies do you have objections? Also looping @RBhupi in as @kelcyno indicated that you have interest in doing storm-level tracking

JuliaKukulies replied:
No objections

kelcyno replied:
No objections here - I'll remove this level in the hierarchy.

freemansw1 commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,477 @@
+    track_parent_storm_id = np.repeat(
+        0, len(track_id)
+    )  # This will always be zero when we don't track larger storm systems *yet*
+    print("found track parent storm ids")
```
I think generally we should avoid debug statements in code, but if we have them, we should use `logging.debug` as in the other modules.

freemansw1 commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,477 @@
+    cell_parent_track_id = list(flatten(cell_parent_track_id))
+    print("found cell parent track ids")
+
+    feature_parent_cell_id = list(TRACK.cell.values.astype(float))
```
Note that these should now be ints with #74

freemansw1 commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,477 @@
+    storm_child_cell_count = [len(cell_id)]
+    storm_child_feature_count = [len(feature_id)]
+
+    storm_dim = "nstorms"
```
I think these dimension names should just be `track`, `cell`, etc.

freemansw1 commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,477 @@
+            "storm_id": (storm_dim, storm_id),
+            "track_id": (track_dim, track_ids),
+            "track_parent_storm_id": (track_dim, track_parent_storm_id),
+            "cell_id": (cell_dim, cell_id),
```
I think we should keep the names as just `cell`, `track`, etc. @JuliaKukulies what do you think?

JuliaKukulies replied:
Yes, agree with this

freemansw1 reviewed (changes requested):
Thanks for these really important changes, @kelcyno. I've made a number of comments throughout, but as I stated today, I'm happy to help you work on any of them. 

I think the biggest thing for me is documentation. It would be great to have more comments throughout the code and an example or two on how to use it. @fsenf do you think the better place for usage examples are here or in the tutorials repo?

freemansw1 commented:
I think given the number of things left to resolve, I would be inclined to push this to v1.5? That would give us documentation and spectral filtering in v1.4. What do you think @JuliaKukulies @w-k-jones @fsenf

JuliaKukulies commented:
> 

That sounds like a good idea if @kelcyno is fine with that as well. Thanks for the additional comments @freemansw1!

kelcyno commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,477 @@
+    if not None in Projection:
+        ds["ProjectionCoordinateSystem"] = Projection
+
+    # Convert the cell ID variable from float to integer
```
Is this referring to the float/integer portion?

kelcyno commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,477 @@
+import xarray as xr
+
+
+def compress_all(nc_grids, min_dims=2):
```
Added compress_all to utils.

kelcyno commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,477 @@
+    return nc_grids
+
+
+def standardize_track_dataset(TrackedFeatures, Mask, Projection=None):
```
Moved standardize_track_dataset to utils.

kelcyno commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,477 @@
+    """
+    Combine a feature mask with the feature data table into a common dataset.
+
+    returned by tobac.themes.tobac_v1.segmentation
```
Updated for tobac v1.x

kelcyno commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,477 @@
+
+    Parameters
+    ----------
+    TRACK : xarray.core.dataset.Dataset
```
I wrapped this so the input is in pandas, and immediately converts to xarray. Same in reverse for the output.

freemansw1 commented:
@kelcyno Are you ready for a re-review on this, or are you still iterating? Sorry, I've been busy and lost track of this.

freemansw1 commented:
Just to note where we are, I think @JuliaKukulies and I are waiting for a few last changes and the documentation file before we re-review, right @kelcyno ?

kelcyno commented:
@freemansw1 @JuliaKukulies I've added documentation and recent changes we discussed in the last tobac meeting. Ready for another round of reviews!

freemansw1 commented:
@kelcyno can you do me a favor and merge the latest `RC_v1.4.0` changes into this PR? I can also do it from my side, but it's probably easier for you to do it from yours.

freemansw1 commented:
Never mind, I figured it out and it went through really quickly.

freemansw1 commented on tobac/utils.py:
```diff
@@ -894,3 +894,174 @@ def spectral_filtering(
+    return nc_grids
+
+
+def standardize_track_dataset(TrackedFeatures, Mask, Projection=None):
```
I have mixed feelings about this function. On the one hand, I think this will help our users as we complete the (admittedly slow) transition to v2.0. On the other hand, I worry that it commits us to a data structure that we haven't really had a chance to discuss or experiment with. Maybe the way to go is to just label this as experimental and essentially state that we don't commit to keeping this data structure with v2.0? I'm not sure.

JuliaKukulies replied:
I was also reflecting a bit on this and I like Sean's idea with labeling it as experimental to make sure we do not commit to this

freemansw1 commented on tobac/utils.py:
```diff
@@ -894,3 +894,174 @@ def spectral_filtering(
         return filtered_field
+
+
+def compress_all(nc_grids, min_dims=2, comp_level=4):
```
I'm not sure that this is needed in tobac, but I also don't have opposition to including it. It feels a bit too specific and not connected enough, but on the other hand it's well documented and solves a relatively common problem.

freemansw1 commented on doc/merge_split.rst:
```diff
@@ -0,0 +1,19 @@
+This submodule will label merged/split cells with a TRACK number in addition to its CELL number.
+
+Features, cells, and tracks are combined using parent/child nomenclature. 
+(quick note on terms; “feature” is a detected object at a single time step. “cell” is a series of features linked together over multiple timesteps. "track" may be an individual cell or series of cells which have merged and/or split.)
```
I think a link here to the feature detection and/or tracking steps would be useful.

freemansw1 commented on doc/merge_split.rst:
```diff
@@ -0,0 +1,19 @@
+
+Features, cells, and tracks are combined using parent/child nomenclature. 
+(quick note on terms; “feature” is a detected object at a single time step. “cell” is a series of features linked together over multiple timesteps. "track" may be an individual cell or series of cells which have merged and/or split.)
+
```
On the one hand, while I think the documentation to the function should not include references to the MST, I think this (or a second page) would be a good place to document how the mergers/splits function actually works, below the output dataframe.

freemansw1 commented on build/lib/tobac/merge_split.py:
```diff
@@ -0,0 +1,259 @@
+"""
```
Actually, this is what is causing the Docker CI to fail.

freemansw1 reviewed (changes requested):
I'm getting happy with this, but I have some comments, primarily around documentation and comments. This will be an excellent capability for tobac to have long-term, and I'm hoping to position it such that others quickly use it and it is easy to maintain for the long-haul.

JuliaKukulies commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,260 @@
+"""
+
+
+def merge_split(TRACK, distance=25000, frame_len=5, dxy=500):
```
Good point, I agree with this. Maybe simply `merge_split_cells` or maybe you have a nicer idea :)

JuliaKukulies commented on doc/merge_split.rst:
```diff
@@ -0,0 +1,19 @@
+
+Overview of the output dataframe from merge_split
+  - cell_parent_track_id: The associated track id for each cell. All cells that have merged or split will have the same parent track id. If a cell never merges/splits, only one cell will have a particular track id. 
+  - feature_parent_cell_id: The associated parent cell id for each feature. All feature in a given cell will have the same cell id. 
```
That is just a tiny thing, but maybe make a table like we did for the output of the feature and tracking dataframes  to be consistent (https://tobac--150.org.readthedocs.build/en/150/feature_detection_output.html)

JuliaKukulies commented on doc/merge_split.rst:
```diff
@@ -0,0 +1,19 @@
+*Merge and Split 
```
We need to refer to this file in `index.rst` to include the documentation on our page

snilsn commented:
Great contribution @kelcyno ! I think this is a really interesting functionality and that's why I wanted to try it.
This is the code to create the small dataset with a merge, which might be a good fit for a test:

```
import pandas as pd
import tobac
import tobac.testing

cell_1 = tobac.testing.generate_single_feature(
        1,
        1,
        min_h1=0,
        max_h1=100,
        min_h2=0,
        max_h2=100,
        frame_start=0,
        num_frames=2,
        spd_h1=20,
        spd_h2=20,
    )
cell_1['feature'] = [1, 3]

cell_2 = tobac.testing.generate_single_feature(
        1,
        100,
        min_h1=0,
        max_h1=100,
        min_h2=0,
        max_h2=100,
        frame_start=0,
        num_frames=2,
        spd_h1=20,
        spd_h2=-20,
    )
cell_2['feature'] = [2, 4]
cell_3 = tobac.testing.generate_single_feature(
        30,
        50,
        min_h1=0,
        max_h1=100,
        min_h2=0,
        max_h2=100,
        frame_start=2,
        num_frames=2,
        spd_h1=20,
        spd_h2=0,
    )
cell_3['feature'] = [5, 6]
features = pd.concat([cell_1, cell_2, cell_3])

output = tobac.linking_trackpy(
        features, None, 1, 1, v_max=30
    )
```
![merge](https://user-images.githubusercontent.com/76663232/194621921-430e4bae-40eb-4572-844f-9fa105719c36.png)

I tried to keep the feature ids matching the frames, but I don't know if that's important. My idea would now be to infer the output that `merge_split_cells()` should return from the dataset and the picture,
put that expected output into the format

```
    expected_d = xr.Dataset(
        {
            "track": (track_dim, track_ids),
            "cell": (cell_dim, cell_id),
            "cell_parent_track_id": (cell_dim, cell_parent_track_id),
            "feature": (feature_dim, feature_id),
            "feature_parent_cell_id": (feature_dim, feature_parent_cell_id),
            "feature_parent_track_id": (feature_dim, feature_parent_track_id),
            "track_child_cell_count": (track_dim, track_child_cell_count),
            "cell_child_feature_count": (cell_dim, cell_child_feature_count),
        }
    )
```
and assert the real output in a test function.

Unfortunately, I can't get past the assert statements at the end of `merge_split_cells()`, which could mean that something is not working properly or that I have gotten something wrong and the dataset needs to be be altered. In either case, I would be happy to contribute further.

kelcyno commented:
@snilsn I did not update the assertion at the end when I updated how I was numbering tracks. I've tried it with your cells and your test dataset made it pass all assert statements. Do you mind continuing to work on the testing aspect? Should we move the assert statements from the end of merge_split_cells? I believe @freemansw1 mentioned at some point that this might be the correct way to handle testing.

freemansw1 commented:
I think the approach that @snilsn has for testing is probably the way to go, and is an excellent contribution to this. I believe using his tests, you can remove the `assert` statements out of the main function and instead do all testing in the testing code.

snilsn commented:
@kelcyno You are right, it works, no idea what happened a week ago.

Now that I've seen the output, I'm not sure if I could tell if this is the right result (e.g. the track id is -1, not sure what that means). I think it would be easier for you to finish the test, since you know what the result should be. It would take a while for me to get used to this parent/child notation and I have almost zero experience with xarray.

JuliaKukulies commented:
Hi @kelcyno! Thanks a lot for updating the documentation page. I committed some minor changes to your `merge_split` branch:
 1. I removed the text you had accidently copied in twice in `merge_split.rst`, fixed some formatting issues in the .rst file and removed the redundant `merge_split_output_vars.rst`
 2. I also solved the merge conflict (caused by the utils module) by pulling in the latest changes from `RC_v1.4.0` 

I hope it was OK that I directly committed the changes, please let me know if you are not happy with them. I am not sure though why they do not appear here as an update of the PR? Do you have the possibility to update the PR? 

Regarding the test: I think we work on this collaboratively given that you were not sure about how to add the unit tests etc. I am also happy to help with this and I think @snilsn already provided the main code needed. But as @snilsn says, maybe you can help with describing what we would expect. @snilsn I also tested your example and I since a track ID of `-1` simply means that no cell has been tracked, I am wondering if this has something with your output cells. If I use your test data, `tobac.linking_trackpy()` gives me three individual cell IDs with no feature sharing the same position. This is different from what your figure suggest, because you indicate only two cells (one orange and one blue): 

![merge_split_test_data](https://user-images.githubusercontent.com/44163060/197727357-440ccec7-f5c6-405e-9d62-4d3cd2cc205d.png)

snilsn commented:
@JuliaKukulies Yes, you are right, I did multiple versions of this in a notebook and have probably used the wrong picture.
I increased `v_max` to get only two cells, but that doesn't change the track ID. My next guess would be to increase the `distance` parameter of `merge_split_cells()`. Maybe the test data has an awkward spacing?

JuliaKukulies commented on tobac/merge_split.py:
```diff
@@ -0,0 +1,231 @@
+"""
+
+
+def merge_split_cells(TRACK, dxy, distance=25000, frame_len=5):
```
Should we require `distance` instead of making it optional or maybe make it dependent on `dxy` when the user does not provide it? I think this would make more sense given that this is really dependent on the data you are using and 25000 could be less than `dxy` in some datasets while it may be much higher than `dxy` in other datasets in which case `merge_split_cells()` becomes very expensive.

snilsn replied:
If we make it dependent, a good educated guess would probably be the search radius defined via `v_max` or `d_max` in `linking_trackpy`, since this should already be finetuned with respect to the data.

JuliaKukulies replied:
I thought of `dxy` since it already is an input parameter but your idea is a good one and makes actually more sense. I am not opposed to require `v_max` from the tracking, but if it feels too complicated we could simply make `distance` mandatory. @kelcyno do you have any opinion?

freemansw1 replied:
I think we should make `distance` mandatory. @kelcyno are you ok with this?

JuliaKukulies commented:
That sounds good to me @freemansw1. Thank you for taking initiative on the test!

JuliaKukulies commented:
> 
> On the testing front, I'll take a stab at adding a test (tests?) based on what @snilsn has done above and will PR to @kelcyno's branch. That seems like the best path forward to me? 

If @kelcyno does not have time to review and is OK with this, you could also directly commit your test to this PR like you have done  @freemansw1

kelcyno commented:
@freemansw1 I'm not a huge fan of only making the distance threshold mandatory.  I think it's important to have a default option, as the other functions all have default values if no keyword is given. We can make it clear in the documentation that inputting the users own distance is recommended, but having a value ~ 10x the dxy term as default is what I prefer. You're within your purview to override me here, but I think we can do a lot with documentation here to help the user.

kelcyno commented:
@JuliaKukulies I can review the PR. Real quick, to make sure our labeling is consistent - the Feature numbering begins at 1, the cell numbering begins at 0 or 1? I've been initiating the Track numbering at 0 but I'm wondering if I need to make it start at 1 to be consistent. Thoughts?

freemansw1 commented:
@kelcyno That's a reasonable perspective. I will note that we probably should have fewer default values, but in general, having a default value of `None` (that we set to 10x `dxy` in the code) would be a great compromise. If we decide not to have a default in the future, that's a lot easier to undo than requiring the parameter (see our #125 headache). 

@kelcyno I'm about to open a PR onto your branch with my test (and a few other changes, too).

freemansw1 commented:
OK, I've opened a PR on @kelcyno's fork with a test and some (small) changes otherwise. https://github.com/kelcyno/tobac/pull/1

freemansw1 commented:
> @JuliaKukulies I can review the PR. Real quick, to make sure our labeling is consistent - the Feature numbering begins at 1, the cell numbering begins at 0 or 1? I've been initiating the Track numbering at 0 but I'm wondering if I need to make it start at 1 to be consistent. Thoughts?


Oh yeah this has been a headache for me recently. We are not consistent on our numbering. In my opinion, we should shift numbering to be 0-indexed universally like the rest of the Python world, but that's a different issue and different PR (and probably not worth addressing until v2.0, given that it's kind of a breaking change). 

I'd say for now, stick with the 0 numbering.

freemansw1 commented:
> OK, I've opened a PR on @kelcyno's fork with a test and some (small) changes otherwise. [kelcyno#1](https://github.com/kelcyno/tobac/pull/1)

In my mind, once this is merged, I'll approve merging this into the RC.

JuliaKukulies commented:
> @JuliaKukulies I can review the PR. Real quick, to make sure our labeling is consistent - the Feature numbering begins at 1, the cell numbering begins at 0 or 1? I've been initiating the Track numbering at 0 but I'm wondering if I need to make it start at 1 to be consistent. Thoughts?

Great, thank you for doing this! And I agree with the thoughts of @freemansw1 on the numbering. 




> > OK, I've opened a PR on @kelcyno's fork with a test and some (small) changes otherwise. [kelcyno#1](https://github.com/kelcyno/tobac/pull/1)
> 
> In my mind, once this is merged, I'll approve merging this into the RC.


Fabulous @freemansw1! I just had a quick look at your test and small changes and everything seems reasonable. Yes, please go ahead with the merge and release once you have the approval of @kelcyno

kelcyno commented:
@JuliaKukulies I have to make a quick change to the merge_split.py, if you'll wait for final approval until then I think we should be good. I'm just testing now before I push it the update later this evening and approve @freemansw1 's changes.

freemansw1 commented:
Thanks for pushing on this @kelcyno! Just trying to get the code done and paper in before I start my new position :). I'll be pushing on #127 here soon with the same goal in mind. I'm also working on a new PR for converting between grids between tracking and segmentation.

kelcyno commented:
@freemansw1 @JuliaKukulies, I've updated Sean's PR, our numbering for Tracks begins at 0. I updated the dimension names to be consistent, and similarly in the utils file the standardize_track_dataset function. I commented out the assert tests at the bottom of the merge_split function, but can delete them entirely if needed. I rather like keeping them for future use if needed. 
What will need some attention quickly is how I've used a default value of None in the merge distance, but internally if there is a 'None' then 25*dxy is used for the distance. 

I *think* it's nearly ready for 1.4 - and all checks have passed (for the first time!). Let me know what you think and if there's changes to be made on the minimum distance value part. Thanks all!

JuliaKukulies reviewed (commented):
@kelcyno I have had a look again at your latest changes and the way you handle `distance` now is fine for me. I do have a few remaining remarks that should be quick fixes and two questions. But otherwise, I am really fine to approve this PR from my side. Have you found the reason why `merge_split` does not occur in the auto-generated API reference @freemansw1 ?

freemansw1 commented:
Good catch on the documentation, @JuliaKukulies! I resolved it by adding `merge_split` to the `tobac.rst` file.

JuliaKukulies commented:
Perfect @freemansw1 ! I was looking for a `api.rst` and missed it :)

kelcyno commented:
@freemansw1 I'm checking an error Eric found, it's running on real data now, but we need to hold off to merge until we're satisfied it wasn't just human error.

freemansw1 commented:
Glad to be finding these now! If it is a bug, once you get it fixed, let's add a test for it to make sure we don't accidentally break it again in the future.

freemansw1 commented:
@kelcyno one more general comment... can we rename the `merge_split_cells` method to something a little more specific that is appropriate to describe the method? As we move forward with *tobac*, I think it would be good for us to more generally name things specifically and allow for multiple methods to be added to do the various steps, and this is a good place to start.

kelcyno commented:
> 
@freemansw1  Do you have something in mind? the method I'm using here is the minimum euclidian spanning tree. It's a weighted graph using  the Kruskal algorithm. Merge_split_spanningtree or merge_split_MEST might be more specific, but I'm not sure if that's what you had in mind.

freemansw1 commented:
> > 
> 
> @freemansw1 Do you have something in mind? the method I'm using here is the minimum euclidian spanning tree. It's a weighted graph using the Kruskal algorithm. Merge_split_spanningtree or merge_split_MEST might be more specific, but I'm not sure if that's what you had in mind.

I will defer to you, but maybe something just like `merge_split_MEST` or `merge_split_min_span_tree`? I'm just trying to keep options open in the future for other merge/split algorithms, especially if we add one that includes storm tracks.

kelcyno commented:
@freemansw1 I changed the function name to merge_split_MEST, and updated the documentation. The error Eric found is in how the track ids are numbered after merging. I simplified how the IDs are numbered using list.index, and used the same methods for cell_parent_track_id, feature_parent_track_id etc. This passed the tobac merge_split test, and also passes our  tests using the TRACER data.

freemansw1 reviewed (approved):
I'm now happy with these changes, once the final comment from @JuliaKukulies is addressed.

JuliaKukulies reviewed (approved):
I am also happy with the new changes! Thanks a lot for your hard work on this, @kelcyno and @freemansw1. How great if we can merge this week! 

@kelcyno it would be great if you can quickly clarify the code snippet I was wondering about in `merge_split.py` before @freemansw1 presses the button. I did not really understand what is done here, but it could be that I just misunderstood the code.

freemansw1 commented:
With #199 now merged, I will release v1.4.0 when this PR is merged.

JuliaKukulies commented:
Thanks @kelcyno for the explanation, that makes sense! I agree with you that this should be kept as it is then. It looks like we are ready to merge @freemansw1  !

freemansw1 commented:
Thanks to @kelcyno and @deeplycloudy for their hard work on getting this amazing new capability into *tobac*. Thanks also to @JuliaKukulies for her reviews and to @snilsn for helping with testing.
